File: MVVM/Converters/ThicknessConverter.vb
Imports System.Globalization
Imports System.Windows.Data

Namespace MVVM.Converters

    Public Class ThicknessConverter
        Implements IValueConverter

        Public Function Convert(ByVal Value As Object, ByVal TargetType As Type, ByVal Parameter As Object, ByVal Culture As CultureInfo) As Object Implements IValueConverter.Convert
            Dim V = CType(Value, Double)
            Dim P = 1.0
            If Parameter IsNot Nothing Then
                P = CType(Parameter, Double)
            End If
            Return New Thickness(P * V * Me.Coefficients.Left,
                                 P * V * Me.Coefficients.Top,
                                 P * V * Me.Coefficients.Right,
                                 P * V * Me.Coefficients.Bottom)
        End Function

        Public Function ConvertBack(ByVal Value As Object, ByVal TargetType As Type, ByVal Parameter As Object, ByVal Culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
            Dim Th = CType(Value, Thickness)
            Dim V As Double?

            ' ToDo Correct the double equality checks. (It should somehow just check the significant digits, and not the order of magnitude...)

            If Me.Coefficients.Left <> 0 Then
                V = Th.Left / Me.Coefficients.Left
            End If
            If Me.Coefficients.Top <> 0 Then
                If V.HasValue And V.Value <> Th.Top / Me.Coefficients.Top Then
                    Return 0
                End If
                V = Th.Top / Me.Coefficients.Top
            End If
            If Me.Coefficients.Right <> 0 Then
                If V.HasValue And V.Value <> Th.Right / Me.Coefficients.Right Then
                    Return 0
                End If
                V = Th.Right / Me.Coefficients.Right
            End If
            If Me.Coefficients.Bottom <> 0 Then
                If V.HasValue And V.Value <> Th.Bottom / Me.Coefficients.Bottom Then
                    Return 0
                End If
                V = Th.Bottom / Me.Coefficients.Bottom
            End If
            If Not V.HasValue Then
                Return 0
            End If
            Dim P = 1.0
            If Parameter IsNot Nothing Then
                P = CType(Parameter, Double)
            End If
            Return V.Value * P
        End Function

        Public Property Coefficients As Thickness

    End Class

End Namespace

File: WPF/MVVM/Converters/ThicknessConverter.cs
using System.Windows;
using System;
using System.Globalization;
using System.Windows.Data;

namespace Ks
{
    namespace Ks.Common.MVVM.Converters
    {
        public class ThicknessConverter : IValueConverter
        {
            public object Convert(object Value, Type TargetType, object Parameter, CultureInfo Culture)
            {
                var V = System.Convert.ToDouble(Value);
                var P = 1.0;
                if (Parameter != null)
                    P = System.Convert.ToDouble(Parameter);
                return new Thickness(P * V * this.Coefficients.Left, P * V * this.Coefficients.Top, P * V * this.Coefficients.Right, P * V * this.Coefficients.Bottom);
            }

            public object ConvertBack(object Value, Type TargetType, object Parameter, CultureInfo Culture)
            {
                var Th = (Thickness)Value;
                double? V = null;

                // ToDo Correct the double equality checks. (It should somehow just check the significant digits, and not the order of magnitude...)

                if (this.Coefficients.Left != (double)0)
                    V = Th.Left / this.Coefficients.Left;
                if (this.Coefficients.Top != (double)0)
                {
                    if (V.HasValue & (V.Value != (Th.Top / this.Coefficients.Top)))
                        return 0;
                    V = Th.Top / this.Coefficients.Top;
                }
                if (this.Coefficients.Right != (double)0)
                {
                    if (V.HasValue & (V.Value != (Th.Right / this.Coefficients.Right)))
                        return 0;
                    V = Th.Right / this.Coefficients.Right;
                }
                if (this.Coefficients.Bottom != (double)0)
                {
                    if (V.HasValue & (V.Value != (Th.Bottom / this.Coefficients.Bottom)))
                        return 0;
                    V = Th.Bottom / this.Coefficients.Bottom;
                }
                if (!V.HasValue)
                    return 0;
                var P = 1.0;
                if (Parameter != null)
                    P = System.Convert.ToDouble(Parameter);
                return V.Value * P;
            }

            public Thickness Coefficients { get; set; }
        }
    }
}
